// react-swipr/js/react-swipr-es6.js
import React, { Component } from 'react';

/**
 * React Swipr component
**/

export default class ReactSwipr extends React.Component{

  static propTypes = {
    elementId: React.PropTypes.string.isRequired,
    navigation: React.PropTypes.bool
  }

  componentDidMount() {

    var elementId = this.props.elementId;
    var element = document.getElementById(elementId);

    if(element == null || element == undefined) {
      return;
    } else {
      this.swipe(element);
    }
        
  }

  swipe = (element)=> {

    // initialize the swiper on the element with the defined id
    var swipr = require('swipr');
    this.swipr = swipr(element, this.props);
  }

  componentDidUpdate() {
    this.swipr.setIndex(this.props.index);
  }

  render(){

    var navigation = null

    if(this.props.navigation){

      navigation = (
          <div>
            <span className="swipr_prev">
              <svg width="50" height="50" viewBox="0 0 501.5 501.5">
                <g>
                  <path d="M302.67 90.877l55.77 55.508L254.575 250.75 358.44 355.116l-55.77 55.506L143.56 250.75z" stroke='#7ED321' strokeWidth='2' fill='ff6700'/>
                </g>
              </svg>
            </span>
            <span className="swipr_next">
              <svg width="50" height="50" viewBox="0 0 501.5 501.5">
                <g>
                  <path d="M199.33 410.622l-55.77-55.508L247.425 250.75 143.56 146.384l55.77-55.507L358.44 250.75z" stroke='#7ED321' strokeWidth='2' fill='ff6700'/>
                </g>
              </svg>
            </span>
          </div>
      );
    }

    return (
        <div className="react-swipr">
          <div className="swipr">
            <ul className="swipr_slides" style={{width:"6.1rem"}}>
              {this.props.children}
            </ul>
          </div>
          {navigation}
        </div>
    );
  }
}
